eventual-cassandra/src/main/scala/com/evolutiongaming/kafka/journal/eventual/cassandra/ReplicatedCassandra.scala

```diff
@@ -429,21 +447,29 @@ private[journal] object ReplicatedCassandra {
                                             .some
                                         }
                                       } yield {
-                                        (deleteTo.value, seqNr)
+                                        // prune yet non-deleted part of journal
+                                        (deleteTo.value, headSeqNr)
                                       }
-                                    } else if (deleteTo.value == seqNr) {
-                                      (seqNr, seqNr).pure[F]
+                                    } else if (deleteTo.value == headSeqNr) {
+                                      // all is fully pruned - nothing to do
+                                      (headSeqNr, headSeqNr).pure[F]
                                     } else {
+                                      // somehow `deleteTo` is ahead of head's `seqNr`
+                                      // TODO MR how this can happen? - in `def append` we clamp down `DeleteTo`
```

Comment:
I am not sure, how this can happen - my current guess is that due to some error we [used to?] get to unexpected state, possibly due to original bug, which gets fixed in this PR

Reply:
I see code for the same case in "main" part of `def delete(...)`:
```
val deleteTo0 = journalHead.deleteTo
val seqNr     = journalHead.seqNr
val seqNr1    = deleteTo0.fold { seqNr } { _.value.max(seqNr) }
if (seqNr1 == seqNr) { ... } else {
  update(seqNr1, ...)
}
```

New implementation lost it btw

Reply:
With state `(delete_to: 384, seqNr:574)` and action `DeleteTo(575)`, after stripping down original code we get:
```scala
val deleteTo0 = journalHead.deleteTo // 384
val seqNr     = journalHead.seqNr // 574
val seqNr1    = deleteTo0.fold { seqNr } { _.value.max(seqNr) } // 574 = max(384, 574)
val deleteTo1 = deleteTo.value.min(seqNr1).toDeleteTo // 574 = min(575, 574)

if (seqNr1 != seqNr && !deleteTo0.contains(deleteTo1)) { // 574 != 574 && 384 != 574
  update(seqNr1, deleteTo1) // doesn't happen
}
```

We would update both `seq_nr` and `delete_to` when both:
* head's `seq_nr` is not equal with head's `delete_to` (non-empty)
* and head's `delete_to` is not equal with clamped down requested action's `deleteTo`

---

In new code against the same state, we get:
```scala
val headDeleteTo = journalHead.deleteTo // 384
val headSeqNr    = journalHead.seqNr // 574
val validatedSeqNr: SeqNr = setSeqNr.fold { headSeqNr } { _.max(headSeqNr) } // 575 = max(575, 574)
val validatedDeleteTo = deleteTo.value.min(validatedSeqNr).toDeleteTo // 575 = min(575, 575)

val updateSeqNr = setSeqNr.isDefined // true
val updateDeleteTo = headDeleteTo.forall(_ < validatedDeleteTo) // true = 574 < 575

if (updateSeqNr && updateDeleteTo) { // true && true
  update(validatedSeqNr, validatedDeleteTo)
```

---

In old code there is no way to curry over information that `Delete` action had at least one `Append` action, which was discarded due to optimization

Reply:
Excuse me but your missed the question:

> // somehow `deleteTo` is ahead of head's `seqNr`

> I see code for the same case in "main" part of def delete(...):

> New implementation lost it btw

Thus the case is: initial state - `delete_to = 400` and `seq_nr = 300` when incoming `DeleteTo = 200`, then
```
                              val deleteTo0 = journalHead.deleteTo // 400
                              val seqNr     = journalHead.seqNr    // 300
                              val seqNr1    = deleteTo0.fold { seqNr } { _.value.max(seqNr) } // 400
                              val deleteTo1 = deleteTo // 200
                                .value
                                .min(seqNr1)
                                .toDeleteTo
                                  if (seqNr1 == seqNr) { // 400 != 300, ie false
                                    if (deleteTo0.contains(deleteTo1)) {
                                      ...
                                  } else {
                                     ... // update seq_nr, this is lost in your new implementation 
                                  }
```

Reply:
I created the unit-test-case in `ReplicatedCassandraTest` like:
```scala
    test(s"fix delete_to, $suffix") {
      val id      = "id"
      val key     = Key(id = id, topic = topic0)
      val segment = segmentOfId(key)
      val stateT = journal.delete(
        key = key,
        Partition.min,
        Offset.unsafe(2),
        timestamp = timestamp1,
        deleteTo  = SeqNr.unsafe(200).toDeleteTo,
        origin    = origin.some,
      )

      val initial = State
        .empty
        .copy(
          metaJournal = Map(
            (
              (topic0, segment),
              Map(
                (
                  id,
                  MetaJournalEntry(
                    journalHead = JournalHead(
                      partitionOffset = PartitionOffset(Partition.min, Offset.unsafe(1)),
                      segmentSize     = segmentSize,
                      seqNr           = SeqNr.unsafe(300),
                      deleteTo        = SeqNr.unsafe(400).toDeleteTo.some,
                      expiry          = none,
                      recordId        = none,
                    ),
                    created = timestamp0,
                    updated = timestamp0,
                    origin  = origin.some,
                  ),
                ),
              ),
            ),
          ),
        )

      val expected = State(
        metaJournal = Map(
          (
            (topic0, segment),
            Map(
              (
                id,
                MetaJournalEntry(
                  journalHead = JournalHead(
                    partitionOffset = PartitionOffset(Partition.min, Offset.unsafe(2)),
                    segmentSize     = segmentSize,
                    seqNr           = SeqNr.unsafe(200), // <========= here test fails - actual value is `400`
                    deleteTo        = SeqNr.unsafe(200).toDeleteTo.some,
                    expiry          = none,
                    recordId        = none,
                  ),
                  created = timestamp0,
                  updated = timestamp1,
                  origin  = origin.some,
                ),
              ),
            ),
          ),
        ),
      )

      val actual = stateT.run(initial)
      actual shouldEqual (expected, true).pure[Try]
    }
```

It fails with

Reply:
we we "fix" by advancing `seqNr`... will add test-case to this PR...

Reply:
> If I read it correct, then journal table would be empty and on recover actor would start with seqNr: 300.

the state can be:
```
meta.seq_nr     = 300
meta.delete_to = 400

journal contains events from 0 to 400
```

> Is the setup you describe even possible?

I have no idea how this state is possible but:
1. i don't believe that the fix logic was added just for fun
2. i don't have any guarantee that we don't have such journals
3. the code _now_ does not allow it but could allow _in past_